### PR TITLE
add disable_policy_batching to cisco_bgp

### PR DESCRIPTION
**Pull request by andy91315**
1. pass beaker test 
2. pass static analysis

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -451,6 +451,20 @@ Routing domain confederation AS. Valid values are String, keyword 'default'.
 ##### `confederation_peers`
 AS confederation parameters. Valid values are String, keyword 'default'.
 
+##### `disable_policy_batching` 
+Enable/Disable the batching evaluation of prefix advertisements to all peers. Valid values are String, keyword 'default'.
```
Comment (chrisvanheuveln) [inline README.md]: Isn't this a boolean?

File: README.md [hunk 2]
```diff
@@ -451,6 +451,20 @@ Routing domain confederation AS. Valid values are String, keyword 'default'.
...
+
+##### `disable_policy_batching_prefix`
+Enable/Disable the batching evaluation of prefix advertisements to all peers with prefix list.
```
Comment (chrisvanheuveln) [inline README.md]: `/with prefix list/with a prefix-list/`

File: README.md [hunk 3]
```diff
@@ -451,6 +451,20 @@ Routing domain confederation AS. Valid values are String, keyword 'default'.
...
+
+For example prefix_list:
```
Comment (chrisvanheuveln) [inline README.md]:
```
Enable/Disable the batching evaluation of prefix advertisements to all peers with one or more prefix-lists. The prefix-lists must be defined in the form of an array. Each entry in the array must include an Address-Family Identifier (ipv4 or ipv6) and the name of the prefix-list.

Example array:
\``` ruby
[
  ["ipv4", "my_v4_prefixes"],
  ["ipv6", "my_v6_prefixes"]
]
```

File: lib/puppet/provider/cisco_bgp/nxapi.rb [hunk 4]
```diff
@@ -108,6 +110,7 @@ def self.properties_get(asn, vrfname, instance)
       end
     end
     debug current_state
+    current_state[:disable_policy_batching_prefix] = instance.disable_policy_batching_prefix
```
Comment (chrisvanheuveln) [inline lib/puppet/provider/cisco_bgp/nxapi.rb]: This should be before the debug and should have a comment:

```
-    debug current_state
+    # Nested array handling
+    current_state[:disable_policy_batching_prefix] = instance.disable_policy_batching_prefix
+    debug current_state
```

File: lib/puppet/provider/cisco_bgp/nxapi.rb [hunk 5]
```diff
@@ -209,6 +212,21 @@ def confederation_peers_set
     @bgp_vrf.confederation_peers_set(@property_flush[:confederation_peers])
   end
 
+  def disable_policy_batching_prefix
+    return @property_hash[:disable_policy_batching_prefix] if @resource[:disable_policy_batching_prefix].nil?
+    if @resource[:disable_policy_batching_prefix][0] == :default &&
+       @property_hash[:disable_policy_batching_prefix] == @bgp_vrf.default_disable_policy_batching_prefix
+      return [:default]
```
Comment (chrisvanheuveln) [inline lib/puppet/provider/cisco_bgp/nxapi.rb]: Please wrap the long lines

```
  def disable_policy_batching_prefix
    return @property_hash[:disable_policy_batching_prefix] if
      @resource[:disable_policy_batching_prefix].nil?
    if @resource[:disable_policy_batching_prefix][0] == :default &&
      @property_hash[:disable_policy_batching_prefix] ==
      @bgp_vrf.default_disable_policy_batching_prefix
      return [:default]
```

File: lib/puppet/provider/cisco_bgp/nxapi.rb [hunk 6]
```diff
@@ -209,6 +212,21 @@ def confederation_peers_set
...
+    else
+      @property_hash[:disable_policy_batching_prefix]
+    end
+  end
+
+  def disable_policy_batching_prefix=(should_list)
+    should_list = @bgp_vrf.default_disable_policy_batching_prefix if should_list[0] == :default
```
Comment (chrisvanheuveln) [inline lib/puppet/provider/cisco_bgp/nxapi.rb]:
```
    should_list = @bgp_vrf.default_disable_policy_batching_prefix if
      should_list[0] == :default
```

File: lib/puppet/type/cisco_bgp.rb [hunk 7]
```diff
@@ -278,6 +280,34 @@ def name
...
+    desc 'Enable/Disable the batching evaluation of prefix' \
+         'advertisements to all peers'
+
+    newvalues(:true, :false, :default)
+  end # property disable_policy_batching
+
+  newproperty(:disable_policy_batching_prefix, array_matching: :all) do
+    format = "[['afi string'], ['prefix-list string']]"
+    desc 'Enable/Disable the batching evaluation of prefix' \
+         "advertisements to all peers.Valid values match format #{format}."
```
Comment (chrisvanheuveln) [inline lib/puppet/type/cisco_bgp.rb]: Please update `desc` to match the README comment I added above.

File: tests/beaker_tests/cisco_bgp/test_bgp_title_patterns.rb [hunk 8]
```diff
@@ -199,6 +200,7 @@
 
     # enforce_first_as only in the default_vrf
     expected_default_values.delete('enforce_first_as')
+    expected_default_values.delete('disable_policy_batching')
```
Comment (chrisvanheuveln) [inline tests/beaker_tests/cisco_bgp/test_bgp_title_patterns.rb]: Why is this needed?

File: README.md (hunk 2 above)
Comment (chrisvanheuveln) [inline README.md]: I think the nested array unnecessarily complicates this property. This should really be two properties:

`disable_policy_batching_ipv4` and `disable_policy_batching_ipv6`, which simply take the prefix-list name string as inputs.

File: tests/beaker_tests/cisco_bgp/test_bgp_title_patterns.rb (hunk 8 above)
Comment (andy91315) [inline tests/beaker_tests/cisco_bgp/test_bgp_title_patterns.rb]: because it is only in the default vrf.

Comment (chrisvanheuveln) [inline tests/beaker_tests/cisco_bgp/test_bgp_title_patterns.rb]: I figured. In that case can you update the comment to:

```
-    # enforce_first_as only in the default_vrf
+    # Remove properties that can only be used in the default vrf
```

File: examples/demo_bgp.pp [hunk 9]
```diff
@@ -19,6 +19,7 @@
   # --------------------------------------------------------------------------#
   # Configure Global BGP                                                      #
   # --------------------------------------------------------------------------#
+  $prefix_list = [['ipv4', 'xx'], ['ipv6', 'yy']]
```
Comment (chrisvanheuveln) [inline examples/demo_bgp.pp]: remove

File: examples/demo_bgp.pp [hunk 10]
```diff
@@ -27,6 +28,8 @@
     cluster_id                             => '55',
     confederation_id                       => '33',
     confederation_peers                    => '99 88 200.1',
+    disable_policy_batching                => true,
+    disable_policy_batching_prefix         => $prefix_list,
```
Comment (chrisvanheuveln) [inline examples/demo_bgp.pp]:
```
-    disable_policy_batching_prefix         => $prefix_list,
+    disable_policy_batching_prefix_ipv4    => 'my_v4_pfx_list'
+    disable_policy_batching_prefix_ipv6    => 'my_v6_pfx_list'
```

File: lib/puppet/type/cisco_bgp.rb [hunk 11]
```diff
@@ -278,6 +280,43 @@ def name
...
+  newproperty(:disable_policy_batching) do
+    desc 'Enable/Disable the batching evaluation of prefix' \
+         'advertisements to all peers'
+
+    newvalues(:true, :false, :default)
+  end # property disable_policy_batching
+
+  newproperty(:disable_policy_batching_ipv4) do
+    desc "Enable/Disable the batching evaluation of prefix
+          advertisements to all peers. Valid values are a string"
```
Comment (chrisvanheuveln) [inline lib/puppet/type/cisco_bgp.rb]: `/a string/String/`

ditto for fail message and for ipv6's

File: tests/beaker_tests/cisco_bgp/test_bgp_defaults.rb [hunk 12]
```diff
@@ -170,6 +171,7 @@
 
   # enforce_first_as only in default_vrf
```
Comment (chrisvanheuveln) [inline tests/beaker_tests/cisco_bgp/test_bgp_defaults.rb]: Please update this comment as mentioned in my previous comment:

```
-    # enforce_first_as only in the default_vrf
+    # Remove properties that can only be used in the default vrf
```

**Comment by chrisvanheuveln** [comment]
Just minor changes still remaining. :+1: after those are fixed.
